python/packages/core/agent_framework/_workflows/_group_chat.py

```diff
@@ -0,0 +1,1200 @@
+    finish: bool = False
+    final_message: ChatMessage | None = None
+
+
+# endregion
+
+
+# region Manager callable
+
+
+GroupChatStateSnapshot = Mapping[str, Any]
+GroupChatManagerFn = Callable[[GroupChatStateSnapshot], Awaitable[GroupChatDirective]]
+
+
+@dataclass
+class GroupChatParticipantSpec:
+    """Metadata describing a single participant in the orchestration.
+
+    Attributes:
+        name: Unique identifier for the participant used by the manager for selection
+        participant: AgentProtocol or Executor instance representing the participant
+        description: Human-readable description provided to the manager for selection context
+    """
+
+    name: str
+    participant: AgentProtocol | Executor
+    description: str
+
+
+GroupChatParticipantPipeline: TypeAlias = Sequence[Executor]
+
+
+@dataclass
+class GroupChatWiring:
+    """Configuration passed to factories during workflow assembly.
+
+    Attributes:
+        manager: Manager instance responsible for orchestration decisions (None when custom factory handles it)
+        manager_name: Display name for the manager in conversation history
+        participants: Mapping of participant names to their specifications
+        max_rounds: Optional limit on manager selection rounds to prevent infinite loops
+        orchestrator: Orchestrator executor instance (populated during build)
+    """
+
+    manager: GroupChatManagerFn | None
+    manager_name: str
+    participants: Mapping[str, GroupChatParticipantSpec]
+    max_rounds: int | None = None
+    orchestrator: Executor | None = None
+
+
+# endregion
+
+
+# region Default participant factory
+
+
+def _default_participant_factory(
+    spec: GroupChatParticipantSpec,
+    _: GroupChatWiring,
+) -> GroupChatParticipantPipeline:
+    """Default factory for constructing participant pipeline nodes in the workflow graph.
+
+    Creates a single AgentExecutor node for AgentProtocol participants or a passthrough executor
+    for custom participants. Translation between group-chat envelopes and the agent runtime is now
+    handled inside the orchestrator, removing the need for dedicated ingress/egress adapters.
+
+    Args:
+        spec: Participant specification containing name, instance, and description
+        _: GroupChatWiring configuration (unused by default implementation)
+
+    Returns:
+        Sequence of executors representing the participant pipeline in execution order
+
+    Behavior:
+        - AgentProtocol participants are wrapped in AgentExecutor with deterministic IDs
+        - Executor participants are wired directly without additional adapters
+    """
+    participant = spec.participant
+    if isinstance(participant, Executor):
+        return (participant,)
+
+    agent = participant
+    agent_executor = AgentExecutor(agent, id=f"groupchat_agent:{spec.name}")
+    return (agent_executor,)
+
+
+# endregion
+
+
+# region Default orchestrator
+
+
+class GroupChatOrchestratorExecutor(Executor):
```

Comment:
All orchestration patterns can be implemented using GroupChat, including Handoff.